modules/internal/BytesStringCommon.chpl

```diff
@@ -0,0 +1,406 @@
+    if t==_bytes then return Bytes.idxType;
+    else if t==string then return String.byteIndex;
+    else compilerError("This function should only be used by bytes or string");
+  }
+
+  proc do_replace(const ref x: ?t, needle: t, replacement: t,
+                  count: int = -1): t where isBytesOrStringType(t) {
+    type _idxt = getIndexType(t);
+    var result: t = x;
+    var found: int = 0;
+    var startIdx: _idxt = 1;
+    const localNeedle: t = needle.localize();
+    const localReplacement: t = replacement.localize();
+
+    while (count < 0) || (found < count) {
+      const idx = result.find(localNeedle, startIdx..);
+      if !idx then break;
+
+      found += 1;
+      result = result[..idx-1] + localReplacement +
+               result[(idx + localNeedle.numBytes)..];
+      var tmp_res = result[..idx-1] + localReplacement +
+               result[(idx + localNeedle.numBytes)..];
+
+      startIdx = idx + localReplacement.numBytes;
+    }
+    return result;
+  }
+
+  iter do_split(const ref x: ?t, sep: t, maxsplit: int = -1,
+                ignoreEmpty: bool = false): t {
+    type _idxt = getIndexType(t);
+
+      if !(maxsplit == 0 && ignoreEmpty && x.isEmpty()) {
+        const localThis: t = x.localize();
+        const localSep: t = sep.localize();
+
+        // really should be <, but we need to avoid returns and extra yields so
+        // the iterator gets inlined
+        var splitAll: bool = maxsplit <= 0;
+        var splitCount: int = 0;
+
+        var start: _idxt = 1;
+        var done: bool = false;
+        while !done  {
+          var chunk: t;
+          var end: _idxt;
+
+          if (maxsplit == 0) {
+            chunk = localThis;
+            done = true;
+          } else {
+            if (splitAll || splitCount < maxsplit) then
+              end = localThis.find(localSep, start..);
+
+            if(end == 0) {
+              // Separator not found
+              chunk = localThis[start..];
+              done = true;
+            } else {
+              chunk = localThis[start..end-1];
+            }
+          }
+
+          if !(ignoreEmpty && chunk.isEmpty()) {
+            // Putting the yield inside the if prevents us from being inlined
+            // in the zippered case, but I don't think there is any way to avoid
+            // that easily
+            yield chunk;
+            splitCount += 1;
+          }
+          start = end+localSep.numBytes;
+        }
+      }
+  }
+
+  // TODO: could use a multi-pattern search or some variant when there are
+  // multiple needles. Probably wouldn't be worth the overhead for small
+  // needles though
+  pragma "no doc"
+  inline proc _startsEndsWith(const ref x: ?t, needles,
+                              param fromLeft: bool) : bool 
+                              where isBytesOrStringType(t) && 
+                                    isHomogeneousTuple(needles) &&
+                                    needles[1].type==t {
```

Comment:
I can do that.

I am curious though, what is the motivation for choosing an if over a where clause? It is obvious that you cannot use an `if` for overload resolution. But for cases such as these, is there a particular drawback for having a where clause? Compiler performance maybe? Or is it only a stylistic concern? If it is only a matter of style, I slightly prefer `where` over `if` here, just because the types that this function can be called with is right in the function declaration (I am using the term "declaration" a bit loosely here).

Reply:
I think a compiler error like `_startsEndsWith does not support this case` is easier to debug than a failure to resolve a call to `_startsEndsWith`.

Reply:
Good point.